provider/vsphere/internal/vsphereclient/client.go

```diff
@@ -441,38 +443,94 @@ func (c *Client) DeleteDatastoreFile(ctx context.Context, datastorePath string)
 	return nil
 }
 
-func (c *Client) destroyVM(
-	ctx context.Context,
-	vm *object.VirtualMachine,
-	taskWaiter *taskWaiter,
-) error {
-	task, err := vm.Destroy(ctx)
-	if err != nil {
-		return errors.Trace(err)
-	}
-	_, err = taskWaiter.waitTask(ctx, task, "destroying VM")
-	return errors.Trace(err)
+// rootDiskB converts the root disk constraint (which uses megabytes for its unit) to bytes
+func rootDiskB(diskMiB uint64) uint64 {
+	return diskMiB * 1024 * 1024
+}
+
+// rootDiskKiB converts the root disk constraint (which uses megabytes for its unit) to kibibytes
+func rootDiskKiB(diskMiB uint64) uint64 {
+	return diskMiB * 1024
 }
 
 func (c *Client) cloneVM(
 	ctx context.Context,
+	args CreateVirtualMachineParams,
 	srcVM *object.VirtualMachine,
-	dstName string,
 	vmFolder *object.Folder,
-	taskWaiter *taskWaiter,
 ) (*object.VirtualMachine, error) {
-	task, err := srcVM.Clone(ctx, vmFolder, dstName, types.VirtualMachineCloneSpec{
-		Config:   &types.VirtualMachineConfigSpec{},
-		Location: types.VirtualMachineRelocateSpec{},
+	_, datacenter, err := c.finder(ctx)
+	if err != nil {
+		return nil, err
+	}
+	taskWaiter := &taskWaiter{
+		args.Clock,
+		args.UpdateProgress,
+		args.UpdateProgressInterval,
+	}
+	configSpec := &types.VirtualMachineConfigSpec{
+		Flags: &types.VirtualMachineFlagInfo{},
+	}
+	if args.Constraints.HasCpuCores() {
+		configSpec.NumCPUs = int32(*args.Constraints.CpuCores)
+	}
+	if args.Constraints.HasMem() {
+		configSpec.MemoryMB = int64(*args.Constraints.Mem)
+	}
+	if args.Constraints.HasCpuPower() {
+		cpuPower := int64(*args.Constraints.CpuPower)
+		configSpec.CpuAllocation = &types.ResourceAllocationInfo{
+			Limit:       &cpuPower,
+			Reservation: &cpuPower,
+		}
+	}
+
+	task, err := srcVM.Clone(ctx, vmFolder, args.Name, types.VirtualMachineCloneSpec{
+		Config: configSpec,
+		Location: types.VirtualMachineRelocateSpec{
+			Pool: &args.ResourcePool,
+		},
 	})
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 	info, err := taskWaiter.waitTask(ctx, task, "cloning VM")
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	vm := object.NewVirtualMachine(c.client.Client, info.Result.(types.ManagedObjectReference))
+
+	if args.Constraints.RootDisk != nil {
+		// The user specified a root disk, so extend the VM's
+		// disk before powering the VM on.
+		args.UpdateProgress(fmt.Sprintf(
+			"extending disk to %s",
+			humanize.IBytes(rootDiskB(*args.Constraints.RootDisk)),
+		))
+		if err := c.extendVMRootDisk(
+			ctx, vm, datacenter,
+			*args.Constraints.RootDisk,
+			taskWaiter,
+		); err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+
+	vmConfigSpec := types.VirtualMachineConfigSpec{}
+	err = c.buildConfigSpec(ctx, args, &vmConfigSpec)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	task, err = vm.Reconfigure(ctx, vmConfigSpec)
```

Comment:
Aren't you setting all of the same values in the config spec that you pass to `srcVM.Clone`? It seems like you'd only need one or the other. I'd call `buildConfigSpec` to build the spec on line 471.

Reply:
Good point. This second config spec isn't needed as we're not dealing with temporary VMs and root disks.